idseq_dag/util/s3.py

```diff
@@ -140,9 +152,22 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
         # This check is a bit imperfect when untarring... unless you follow the discipline that
         # all contents of file foo.tar are under directory foo/...
         if os.path.exists(dst):
-            # No need to fetch this file from s3, it has been just produced
-            # on this instance.
-            return dst
+            # Destination filename exists. If it is in the reference download directory, check the reference fetch log
+            # for a record of the source key and etag, and short-circuit the download if they match. Otherwise, blindly
+            # assume that it's the same file and short-circuit the download.
+            if abspath.startswith(config["REF_DIR"]):
+                try:
+                    with open(os.path.join(config["REF_FETCH_LOG_DIR"], abspath_hash)) as fh:
+                        fetch_record = json.load(fh)
+                    obj = s3.Bucket(parsed_s3_url.netloc).Object(parsed_s3_url.path.lstrip('/'))
+                    assert fetch_record["bucket_name"] == obj.bucket_name
+                    assert fetch_record["key"] == obj.key
+                    assert fetch_record["e_tag"] == obj.e_tag
+                    return dst
+                except Exception:
+                    pass
```

Comment:
Shouldn't this "rm -rf {dst}" instead of "pass"?

It's a minor concern, but the fall-through path previously only ran when "dst" didn't exist, so it may be better to ensure that.

In practice this can be noticed only when automatically untarrig "foo.tar.gz" into "foo/..." and without the deleting we would see some files from the previous download even if the file is downloaded and untarred again on top.

Reply:
I consider it better/more minimalistic to let the following command overwrite the file.

Reply:
It may be not just a single file but a subdirectory untarred.   In which case it wouldn't delete any files left over that shouldn't be present in the new version.